### PR TITLE
[wapi] Fix timed wait code to not overflow. Fixes #42688

### PR DESCRIPTION
**Pull request by kumpera**
Wait math must always subtract ticks and get deltas. That way we avoid overflows.

### CODE REVIEW

**Comment by kumpera** [comment]
@rolfbjarne does this need to land on C7?

File: mono/io-layer/wait.c [hunk 1]
```diff
@@ -139,8 +139,10 @@ guint32 WaitForSingleObjectEx(gpointer handle, guint32 timeout,
 		goto done;
 	}
 	
-	if (timeout != INFINITE)
-		end = mono_100ns_ticks () + timeout * 1000 * 10;
+	if (timeout != INFINITE) {
+		wait_start = mono_100ns_ticks ();
+		timeout_in_ticks = (gint64)timeout * 10 * 1000; //can't overflow as timeout is 32bits
```
Comment (rolfbjarne) [inline mono/io-layer/wait.c]: Is this equivalent to `((gint64)timeout) * 10 * 1000` or ` (gint64)(timeout * 10 * 1000)`?

**Comment by rolfbjarne** [comment]
It's a bad regression, and it's in C7, so I think it should get backported.

File: mono/io-layer/wait.c (hunk 1 above)
Comment (kumpera) [inline mono/io-layer/wait.c]: It's ((guint64)timeout) * 10 * 1000.
I though regular associativity rules would be clear here.

**Comment by kumpera** [comment]
Failures are unrelated. They happen on master for different reasons.

**Pull request merged**